pkg/engine/lifeycletest/test_plan.go

```diff
@@ -0,0 +1,329 @@
+package lifecycletest
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/mitchellh/copystructure"
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/pulumi/pulumi/pkg/v2/engine"
+	"github.com/pulumi/pulumi/pkg/v2/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v2/resource/deploy/providers"
+	"github.com/pulumi/pulumi/pkg/v2/util/cancel"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/result"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/workspace"
+)
+
+type updateInfo struct {
+	project workspace.Project
+	target  deploy.Target
+}
+
+func (u *updateInfo) GetRoot() string {
+	return ""
+}
+
+func (u *updateInfo) GetProject() *workspace.Project {
+	return &u.project
+}
+
+func (u *updateInfo) GetTarget() *deploy.Target {
+	return &u.target
+}
+
+type TestOp func(UpdateInfo, *Context, UpdateOptions, bool) (ResourceChanges, result.Result)
+type ValidateFunc func(project workspace.Project, target deploy.Target, entries JournalEntries,
+	events []Event, res result.Result) result.Result
+
+func (op TestOp) Run(project workspace.Project, target deploy.Target, opts UpdateOptions,
+	dryRun bool, backendClient deploy.BackendClient, validate ValidateFunc) (*deploy.Snapshot, result.Result) {
+
+	return op.RunWithContext(context.Background(), project, target, opts, dryRun, backendClient, validate)
+}
+
+func (op TestOp) RunWithContext(
+	callerCtx context.Context, project workspace.Project,
+	target deploy.Target, opts UpdateOptions, dryRun bool,
+	backendClient deploy.BackendClient, validate ValidateFunc) (*deploy.Snapshot, result.Result) {
+
+	// Create an appropriate update info and context.
+	info := &updateInfo{project: project, target: target}
+
+	cancelCtx, cancelSrc := cancel.NewContext(context.Background())
+	done := make(chan bool)
+	defer close(done)
+	go func() {
+		select {
+		case <-callerCtx.Done():
+			cancelSrc.Cancel()
+		case <-done:
+		}
+	}()
+
+	events := make(chan Event)
```

Comment:
Is this channel closed elsewhere?

Reply:
I believe so. I'll double-check after the fact, as I need to land this PR to unblock some refactoring